sdk/typescript/dev/src/index.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  dag,
+  Container,
+  object,
+  func,
+  Directory,
+  field,
+} from "@dagger.io/dagger"
+
+@object()
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+class TypescriptSdkDev {
+  /**
+   * Project dev environment.
+   * Open a shell in with: `dagger call --source=. project terminal --cmd=/bin/sh`
+   */
+  @field()
+  project: Container
+
+  constructor(source: Directory) {
+    // Extract package.json and yarn.lock to a temporary directory
+    const dependencyFiles = dag
+      .directory()
+      .withFile("package.json", source.file("package.json"))
+      .withFile("yarn.lock", source.file("yarn.lock"))
+
+    // Get source without generated files nor useless files.
+    const sourceCode = dag.directory().withDirectory("/", source, {
+      include: [
+        "**/*.ts",
+        "tsconfig.json",
+        "package.json",
+        "yarn.lock",
+        ".mocharc.json",
+        ".eslintrc.cjs",
+        ".prettierrc.cjs",
+      ],
+      exclude: ["node_modules", "dist", "dev"],
+    })
+
+    // Install dependencies and add source code.
+    this.project = dag
+      .node()
+      .withPkgManager("yarn")
+      .withSource(dependencyFiles)
+      .install() // Install dependencies prior to adding source to improve caching
+      .withSource(sourceCode)
+      .container()
```

Comment:
The way I do this in python is to have a `base()` function that other methods reuse, instead of doing it in the constructor. Not saying it's better, just another option.

Reply:
I like the constructor option, but I might refactor it later to be more flexible around the source provided to the module.